src/Cli/dotnet/Installer/Windows/PipeStreamMessageDispatcherBase.cs

```diff
@@ -0,0 +1,118 @@
+    /// </summary>
+    [SupportedOSPlatform("windows")]
+    internal class PipeStreamMessageDispatcherBase
+    {
+        /// <summary>
+        /// The maxmimum length of a message.
+        /// </summary>
+        public const int MaxMessageSize = 2048;
+
+        /// <summary>
+        /// The backing stream used for reading & writing messages.
+        /// </summary>
+        private PipeStream _pipeStream;
+
+        /// <summary>
+        /// The number of milliseconds to wait for a pipe connection to be established. See <see cref="Connect"/>.
+        /// </summary>
+        public const int ConnectionTimeout = 30000;
+
+        /// <summary>
+        /// Returns <see langword="true"/> if the dispatcher is using a <see cref="NamedPipeClientStream"/>.
+        /// </summary>
+        public bool IsClient => _pipeStream is NamedPipeServerStream;
+
+        /// <summary>
+        /// Gets whether the underlying stream is connected.
+        /// </summary>
+        public bool IsConnected => _pipeStream.IsConnected;
+       
+        /// <summary>
+        /// Creates a new <see cref="PipeStreamMessageDispatcherBase"/> instance.
+        /// </summary>
+        /// <param name="pipeStream">The pipe stream to use for reading and writing messages. The pipe must be configured
+        /// to use <see cref="PipeTransmissionMode.Message"/>.</param>
+        /// <exception cref="ArgumentNullException" />
+        public PipeStreamMessageDispatcherBase(PipeStream pipeStream)
+        {
+            _pipeStream = pipeStream ?? throw new ArgumentNullException(nameof(pipeStream));
+        }
+
+        /// <summary>
+        /// Waits for the underlying <see cref="PipeStream"/> to establish a connection. If the stream is a 
+        /// <see cref="NamedPipeClientStream"/>, its transmission mode will be set to <see cref="PipeTransmissionMode.Message"/>
+        /// once the connection is established.
+        /// </summary>
+        /// <exception cref="TimeoutException"/>
+        public void Connect()
+        {
+            if (!_pipeStream.IsConnected)
+            {
+                if (_pipeStream is NamedPipeServerStream)
+                {
+                    Task connectTask = Task.Factory.StartNew(() => ((NamedPipeServerStream)_pipeStream).WaitForConnection());
+                    connectTask.Wait(ConnectionTimeout);
+
+                    if (!connectTask.IsCompleted)
+                    {
+                        throw new TimeoutException("Timed out waiting for connection.");
+                    }
+                }
+                else if (_pipeStream is NamedPipeClientStream)
+                {
+                    ((NamedPipeClientStream)_pipeStream).Connect(ConnectionTimeout);
+                    // We cannot set the transmission mode of a pipe client until it's connected.
+                    _pipeStream.ReadMode = PipeTransmissionMode.Message;
+                }
```

Comment:
Else throw? Or is there a case where the pipe stream is neither the client nor the server?